src/transactions/InvokeHostFunctionOpFrame.cpp

```diff
@@ -380,69 +380,69 @@ InvokeHostFunctionOpFrame::doApply(Application& app, AbstractLedgerTxn& ltx,
-                    releaseAssertOrThrow(ttlLtxe);
                     if (!isLive(ttlLtxe.current(), ltx.getHeader().ledgerSeq))
                     {
-                        if (isTemporaryEntry(lk))
-                        {
-                            // For temporary entries, treat the non-live entry
-                            // as if the key did not exist
-                            shouldAddEntry = false;
-                        }
-                        else
+                        // For temporary entries, treat the expired entry as
+                        // if the key did not exist
+                        if (!isTemporaryEntry(lk))
                         {
                             // Cannot access an archived entry
                             this->innerResult().code(
                                 INVOKE_HOST_FUNCTION_ENTRY_ARCHIVED);
                             return false;
                         }
                     }
-
-                    TTLEntry = ttlLtxe.current().data.ttl();
+                    else
+                    {
+                        sorobanEntryLive = true;
+                        ttlEntry = ttlLtxe.current().data.ttl();
+                    }
                 }
+                // If ttlLtxe doesn't exist, this is a new Soroban entry
+            }
 
-                if (shouldAddEntry)
+            if (!isSorobanEntry(lk) || sorobanEntryLive)
+            {
+                auto ltxe = ltx.loadWithoutRecord(lk);
+                if (ltxe)
                 {
-                    auto leBuf = toCxxBuf(le);
+                    auto leBuf = toCxxBuf(ltxe.current());
+                    entrySize = static_cast<uint32>(leBuf.data->size());
 
-                    // For entry types that don't have an TTLEntry (i.e.
+                    // For entry types that don't have an ttlEntry (i.e.
                     // Accounts), the rust host expects an "empty" CxxBuf such
                     // that the buffer has a non-null pointer that points to an
                     // empty byte vector
                     auto ttlBuf =
-                        TTLEntry
-                            ? toCxxBuf(*TTLEntry)
+                        ttlEntry
+                            ? toCxxBuf(*ttlEntry)
                             : CxxBuf{std::make_unique<std::vector<uint8_t>>()};
 
-                    entrySize = static_cast<uint32>(leBuf.data->size());
-                    if (TTLEntry)
-                    {
-                        entrySize += static_cast<uint32_t>(ttlBuf.data->size());
-                    }
-
                     ledgerEntryCxxBufs.emplace_back(std::move(leBuf));
-                    TTLEntryCxxBufs.emplace_back(std::move(ttlBuf));
+                    ttlEntryCxxBufs.emplace_back(std::move(ttlBuf));
+                }
+                else
+                {
+                    releaseAssertOrThrow(!sorobanEntryLive);
```

Comment:
I think `releaseAssert(!ttlEntry)` would be better here. This doesn't catch the invalid case where there is an expired Temporary TTL Entry with no associated LedgerEntry.

Reply:
Yeah, this check seems a bit confusing, I would also add a `if (isSorobanEntry)` condition on the assertion to not mix soroban and non-soroban cases.

Reply:
Fixed